docs/rfcs/confmap-provider-configuration.md

```diff
@@ -0,0 +1,204 @@
+specifies the format of a URI, specifies the different parts of a URI and
+suggests two places where we could pass options to Providers: queries and
+fragments.
+
+#### Queries
+
+Advantages:
+
+- Explicitly intended to specify non-hierarchical data in a URI.
+- Often used for this purpose.
+- Fits into existing config URIs for URL-based Providers.
+
+Disadvantages:
+
+- Completely consuming the parameters would mean we can't pass them to config
+  backends. The frequent use of query parameters in server endpoints would force
+  us to limit the types of requests we can make.
+- Partially consuming parameters would still cause conflicts if a Provider
+  parameter and config backend parameter share the same name. Additionally,
+  mixing the two would provide suboptimal UX.
+- Only allows easily specifying key-value pairs.
+
+#### Fragments
+
+We could specify options in a query parameter-encoded string placed into the URI
+fragment.
+
+Advantages:
+
+- Not likely to be used by config backends for any of our supported protocols,
+  so has a low chance of conflict.
+- Fits into existing config URIs for URL-based Providers.
+
+Disadvantages:
+
+- Even if fragments are likely not useful to backends, we are still preventing
+  their use in upstream Providers.
+- Doesn't conform to the spirit of how fragments should be used according to RFC
+  3986.
+- Will not work with Providers like the env Provider. This is likely minor, as
+  the env Provider can be expected to work differently than URL-based Providers.
+- Only allows easily specifying key-value pairs.
+
+We could likely partially circumvent the key-value pair limitation by
+recursively calling confmap Providers to resolve files, env vars, HTTP URLs,
+etc. For example:
+
+```text
+https://config.com/config#refresh-interval=env:REFRESH_INTERVAL&headers=file:headers.yaml
+```
+
+Using this strategy would also allow us to more easily get env vars and to get
+values from files for things like API tokens.
+
+### Add to the string passed to the config flag
+
+Instead of configuring Providers inside their URIs, we could add an extra
+section to the string passed to `--config` and configure how a URI is retrieved
+through that config. This approach is undesirable because the config flag
+currently only takes valid URIs, and updating the format to accommodate this
+would require we adopt an unconventional format.
+
+### Separate flags to configure Providers per config URI
+
+Advantages:
+
+- Allows us to keep config URIs opaque.
+- Options live right next to config URIs on the command line.
+- Does not require breaking changes.
+
+Disadvantages:
+
+- The flags would need to be placed in a certain position in the arguments list
+  to specify which URI they apply to.
+- Configuring URIs present in files requires users to look in two places for
+  each URI and is suboptimal UX.
+- Complicating the flags like this would be suboptimal UX.
+
+### Separate config file
```

Comment:
I haven't gotten to read this whole RFC yet, but I very much want to avoid this option as I think it will be pretty confusing.

Reply:
I agree, a config file to configure a config file isn't ideal.